server/src/main/java/org/elasticsearch/env/NodeEnvironment.java

```diff
@@ -1030,13 +1036,44 @@ private static void ensureAtomicMoveSupported(final NodePath[] nodePaths) throws
         }
     }
 
-    /**
-     * Resolve the custom path for a index's shard.
-     * Uses the {@code IndexMetaData.SETTING_DATA_PATH} setting to determine
-     * the root path for the index.
-     *
-     * @param indexSettings settings for the index
-     */
+    private void ensureNoIndexData(final NodePath[] nodePaths) throws IOException {
+        List<Path> badDataPaths = new ArrayList<>();
+        for (NodePath nodePath : nodePaths) {
+            Path indicesPath = nodePath.indicesPath;
+            if (Files.isDirectory(indicesPath)) {
+                try (DirectoryStream<Path> indexStream = Files.newDirectoryStream(indicesPath)) {
+                    for (Path indexPath : indexStream) {
+                        if (Files.isDirectory(indexPath)) {
+                            try (Stream<Path> shardStream = Files.list(indexPath)) {
+                                shardStream.filter(this::isShardPath)
+                                    .forEach(badDataPaths::add);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        if (!badDataPaths.isEmpty()) {
+            throw new IllegalArgumentException("Node is started with "
+                + Node.NODE_DATA_SETTING.getKey()
+                + "=false, but found shard data: "
+                + badDataPaths);
```

Comment:
Is this the absolute path that we're outputting here?

Reply:
This would be the relative path if path.node is relative. Do you prefer the absolute path?

Reply:
Changed it to absolute path.